language/move-binary-format/src/deserializer.rs

```diff
@@ -43,9 +43,21 @@ impl CompiledModule {
         binary: &[u8],
         max_binary_format_version: u32,
     ) -> BinaryLoaderResult<Self> {
-        let module = deserialize_compiled_module(binary, max_binary_format_version)?;
-        BoundsChecker::verify_module(&module)?;
-        Ok(module)
+        let prev_state = move_core_types::state::set_state(VMState::DESERIALIZER);
+        let result = std::panic::catch_unwind(|| {
+            let module = deserialize_compiled_module(binary, max_binary_format_version)?;
+            BoundsChecker::verify_module(&module)?;
+
+            Ok(module)
+        })
+        .unwrap_or_else(|_| {
+            Err(PartialVMError::new(
+                StatusCode::VERIFIER_INVARIANT_VIOLATION,
```

Comment:
I would go for a different and specific error here. `VERIFIER_INVARIANT_VIOLATION` may be handled in different ways from system embedding the VM and  this error really has strong implications.
To be honest I am not even clear `VERIFIER_INVARIANT_VIOLATION` is an "impossible" error.
Thoughts?